chainerrl/agents/trpo.py

```diff
@@ -0,0 +1,572 @@
+            statistics of step sizes of policy updates.
+
+    Statistics:
+        average_value: Average of value predictions on non-terminal states.
+            It's updated before the value function is updated.
+        average_entropy: Average of entropy of action distributions on
+            non-terminal states. It's updated on act_and_train.
+        average_kl: Average of KL divergence between old and new policies.
+            It's updated after the policy is updated.
+        average_policy_step_size: Average of step sizes of policy updates
+            It's updated after the policy is updated.
+    """
+
+    saved_attributes = ['policy', 'vf', 'vf_optimizer', 'obs_normalizer']
+
+    def __init__(self,
+                 policy,
+                 vf,
+                 vf_optimizer,
+                 obs_normalizer=None,
+                 gamma=0.99,
+                 lambd=0.95,
+                 phi=lambda x: x,
+                 entropy_coef=0.01,
+                 update_interval=2048,
+                 max_kl=0.01,
+                 vf_epochs=3,
+                 vf_batch_size=64,
+                 standardize_advantages=True,
+                 line_search_max_backtrack=10,
+                 conjugate_gradient_max_iter=10,
+                 conjugate_gradient_damping=1e-2,
+                 act_deterministically=False,
+                 value_stats_window=1000,
+                 entropy_stats_window=1000,
+                 kl_stats_window=100,
+                 policy_step_size_stats_window=100,
+                 logger=getLogger(__name__),
+                 ):
+
+        if not _is_double_backprop_supported:
+            raise RuntimeError("""\
+You're using Chainer v{}. TRPO requires Chainer v3.0.0 or newer.""".format(chainer.__version__))  # NOQA
+
+        self.policy = policy
+        self.vf = vf
+        self.vf_optimizer = vf_optimizer
+        self.obs_normalizer = obs_normalizer
+        self.gamma = gamma
+        self.lambd = lambd
+        self.phi = phi
+        self.entropy_coef = entropy_coef
+        self.update_interval = update_interval
+        self.max_kl = max_kl
+        self.vf_epochs = vf_epochs
+        self.vf_batch_size = vf_batch_size
+        self.standardize_advantages = standardize_advantages
+        self.line_search_max_backtrack = line_search_max_backtrack
+        self.conjugate_gradient_max_iter = conjugate_gradient_max_iter
+        self.conjugate_gradient_damping = conjugate_gradient_damping
+        self.act_deterministically = act_deterministically
+        self.logger = logger
+
+        self.value_record = collections.deque(maxlen=value_stats_window)
+        self.entropy_record = collections.deque(maxlen=entropy_stats_window)
+        self.kl_record = collections.deque(maxlen=kl_stats_window)
+        self.policy_step_size_record = collections.deque(
+            maxlen=policy_step_size_stats_window)
+
+        self.xp = self.policy.xp
```

Comment:
If it's allowed to put `policy` and `vf` on different devices (cpu / gpu), `self.xp` might be confusing.